In [1]:
import numpy as np 
import urdf2casadi.urdfparser as u2c
from urdf2casadi.geometry import plucker
from urdf_parser_py.urdf import URDF, Pose
import PyKDL as kdl
import kdl_parser_py.urdf as kdlurdf
from timeit import Timer, timeit, repeat
import rbdl
import pybullet as pb


In [2]:
def median(lst):
    n = len(lst)
    if n < 1:
            return None
    if n % 2 == 1:
            return sorted(lst)[n//2]
    else:
            return sum(sorted(lst)[n//2-1:n//2+1])/2.0
        
def average(lst): 
    return sum(lst) / len(lst) 


def c_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToCoriolis(q_kdl, qdot_kdl, g_kdl)


def c_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_u2c(q_none, qdot_none)
    
def c_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    c_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot)
    
    
def c_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.NonlinearEffects(rbdlmodel, q_np, qdot_np, c_rbdl)


def g_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToGravity(q_kdl, g_kdl)


def g_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_u2c(q_none)
    
def g_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot, qddot)
    
def g_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, g_rbdl)
    
    
def ID_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    ID_u2c(q_none, qdot_none, qddot_none)
    
    
def ID_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2 
    ID_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot_none)
    
    
def ID_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, ID_rbdl)

    
def M_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToMass(q_kdl, M_kdl)


def M_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    M_u2c(q_none)
    
def M_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    H = pb.calculateMassMatrix(pbmodel, q_none)
    
    
def M_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.CompositeRigidBodyAlgorithm(rbdlmodel, q_np, M_rbdl)
    
    
def aba_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    aba_u2c(q_none, qdot_none, tau_none)
    
    
def crba_u2c_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    crba_u2c(q_none, qdot_none, tau_none)
    
    
def fd_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.ForwardDynamics(rbdlmodel, q_np, qdot_np, tau_np, qddot_np)
    

In [3]:
sim = pb.connect(pb.DIRECT)
ndofs = 10
urdf_nr = list(range(ndofs + 1))
urdf_nr.pop(0)


nitr = 1000
nrepeat = 100
    

path_to_urdf = "/home/lmjohann/urdf2casadi/examples/urdf/ur5_mod.urdf"
root = 'base_link'
tip = "tool0"

ok, kdl_tree = kdlurdf.treeFromFile(path_to_urdf)
kdlmodel = kdl_tree.getChain(root,tip)

rbdlmodel = rbdl.loadModel(path_to_urdf)
pbmodel = pb.loadURDF(path_to_urdf, useFixedBase=True, flags = pb.URDF_USE_INERTIA_FROM_FILE)
asd = u2c.URDFparser()
robot_desc = asd.from_file(path_to_urdf)

jlist, names, qmax, qmin = asd.get_joint_info(root, tip)
njoints = asd.get_n_joints(root, tip)

gravity_kdl = kdl.Vector()
gravity_kdl[2] = -9.81
gravity_u2c = [0, 0, -9.81]

#rbdl declarations
q_np = np.zeros(njoints)
qdot_np = np.zeros(njoints)
qddot_np = np.zeros(njoints)
g_rbdl = np.zeros(njoints)
c_rbdl = np.zeros(njoints)
ID_rbdl = np.zeros(njoints)

#kdl declatations
q_kdl = kdl.JntArray(njoints)
qdot_kdl = kdl.JntArray(njoints)
g_kdl = kdl.JntArray(njoints)

#u2c and pybullet declarations
q_none = [None]*njoints
qdot_none = [None]*njoints
qddot_none = [None]*njoints
g_pb = [None]*njoints

qddot = []
for i in range(njoints):
    qddot.append(0.)

qdot = []
for i in range(njoints):
    qdot.append(0.)
    
    
M_kdl = kdl.JntSpaceInertiaMatrix(njoints)
M_rbdl = (rbdlmodel.q_size, rbdlmodel.q_size)
M_rbdl = np.zeros(M_rbdl)
tau_np = np.zeros(njoints)
tau_none = [None]*njoints


aba_u2c = asd.get_forward_dynamics_aba(root, tip, gravity_u2c)
M_u2c = asd.get_inertia_matrix_crba(root, tip)
ID_u2c = asd.get_inverse_dynamics_rnea(root, tip, gravity_u2c)
c_u2c = asd.get_coriolis_rnea(root, tip)
g_u2c = asd.get_gravity_rnea(root, tip, gravity_u2c)

crba_u2c = asd.get_forward_dynamics_crba(root, tip, gravity_u2c)

The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.
/home/lmjohann/urdf2casadi/examples/urdf/ur5_mod.urdf


In [27]:
timeit_g_u2c = repeat("g_u2c_func()", setup = "from __main__ import g_u2c_func", repeat = nrepeat, number = nitr)
mediantime_g_u2c  = median(timeit_g_u2c)/nitr*1000000

#timeit_g_pb = repeat("g_pb_func()", setup = "from __main__ import g_pb_func", repeat = nrepeat, number = nitr)
#mediantime_g_pb = median(timeit_g_pb)/nitr*1000000

#timeit_g_rbdl = repeat("g_rbdl_func()", setup = "from __main__ import g_rbdl_func", repeat = nrepeat, number = nitr)
#mediantime_g_rbdl = median(timeit_g_rbdl)/nitr*1000000

#timeit_g_kdl = repeat("g_kdl_func()", setup = "from __main__ import g_kdl_func", repeat = nrepeat, number = nitr)
#mediantime_g_kdl = median(timeit_g_kdl)/nitr*1000000


timeit_c_u2c = repeat("c_u2c_func()", setup = "from __main__ import c_u2c_func", repeat = nrepeat, number = nitr)
mediantime_c_u2c  = median(timeit_c_u2c)/nitr*1000000

#timeit_c_pb = repeat("c_pb_func()", setup = "from __main__ import c_pb_func", repeat = nrepeat, number = nitr)
#mediantime_c_pb = median(timeit_c_pb)/nitr*1000000

#timeit_c_rbdl = repeat("c_rbdl_func()", setup = "from __main__ import c_rbdl_func", repeat = nrepeat, number = nitr)
#mediantime_c_rbdl = median(timeit_c_rbdl)/nitr*1000000

#timeit_c_kdl = repeat("c_kdl_func()", setup = "from __main__ import c_kdl_func", repeat = nrepeat, number = nitr)
#mediantime_c_kdl = median(timeit_c_kdl)/nitr*1000000

In [28]:
timeit_id_u2c = repeat("ID_u2c_func()", setup = "from __main__ import ID_u2c_func", repeat = nrepeat, number = nitr)
mediantime_id_u2c  = median(timeit_id_u2c)/nitr*1000000

#timeit_id_pb = repeat("ID_pb_func()", setup = "from __main__ import ID_pb_func", repeat = nrepeat, number = nitr)
#mediantime_id_pb = median(timeit_id_pb)/nitr*1000000

#timeit_id_rbdl = repeat("ID_rbdl_func()", setup = "from __main__ import ID_rbdl_func", repeat = nrepeat, number = nitr)
#mediantime_id_rbdl = median(timeit_id_rbdl)/nitr*1000000

In [29]:
timeit_M_u2c = repeat("M_u2c_func()", setup = "from __main__ import M_u2c_func", repeat = nrepeat, number = nitr)
mediantime_M_u2c  = median(timeit_M_u2c)/nitr*1000000

#timeit_M_pb = repeat("M_pb_func()", setup = "from __main__ import M_pb_func", repeat = nrepeat, number = nitr)
#mediantime_M_pb = median(timeit_M_pb)/nitr*1000000

#timeit_M_rbdl = repeat("M_rbdl_func()", setup = "from __main__ import M_rbdl_func", repeat = nrepeat, number = nitr)
#mediantime_M_rbdl = median(timeit_M_rbdl)/nitr*1000000

#timeit_M_kdl = repeat("M_kdl_func()", setup = "from __main__ import M_kdl_func", repeat = nrepeat, number = nitr)
#mediantime_M_kdl = median(timeit_M_kdl)/nitr*1000000

In [30]:
timeit_fdaba_u2c = repeat("aba_u2c_func()", setup = "from __main__ import aba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdaba_u2c = median(timeit_fdaba_u2c)/nitr*1000000



timeit_fdcrba_u2c = repeat("crba_u2c_func()", setup = "from __main__ import crba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdcrba_u2c = median(timeit_fdcrba_u2c)/nitr*1000000


#timeit_fd_rbdl = repeat("fd_rbdl_func()", setup = "from __main__ import fd_rbdl_func", repeat = nrepeat, number = nitr)
#mediantime_fd_rbdl = median(timeit_fd_rbdl)/nitr*1000000


In [31]:
print mediantime_g_u2c
print mediantime_c_u2c
print mediantime_id_u2c

16.9825553894
27.3809432983
37.9384756088


In [32]:
print "\nur5_g_u2c=", mediantime_g_u2c
print "\nur5_id_u2c=", mediantime_id_u2c
print "\nur5_C_u2c=", mediantime_c_u2c
print "\nur5_M_u2c=", mediantime_M_u2c

print "\nur5_g_rbdl=", mediantime_g_rbdl
print "\nur5_id_rbdl=", mediantime_id_rbdl
print "\nur5_C_rbdl=", mediantime_c_rbdl
print "\nur5_M_rbdl=", mediantime_M_rbdl

print "\nur5_g_pb=", mediantime_g_pb
print "\nur5_id_pb=", mediantime_id_pb
print "\nur5_C_pb=", mediantime_c_pb
print "\nur5_M_pb=", mediantime_M_pb

print "\nur5_g_kdl=", mediantime_g_kdl
print "\nur5_C_kdl=", mediantime_c_kdl
print "\nur5_M_kdl=", mediantime_M_kdl

print "\fdrbdl=", mediantime_fd_rbdl
print "\nfdcrba=", mediantime_fdcrba_u2c 
print "\nfdaba=", mediantime_fdaba_u2c


ur5_g_u2c= 16.9825553894

ur5_id_u2c= 37.9384756088

ur5_C_u2c= 27.3809432983

ur5_M_u2c= 16.5120363235

ur5_g_rbdl= 6.27994537354

ur5_id_rbdl= 12.0394229889

ur5_C_rbdl= 8.66198539734

ur5_M_rbdl= 5.52415847778

ur5_g_pb= 5.00798225403

ur5_id_pb= 10.1770162582

ur5_C_pb= 7.7269077301

ur5_M_pb= 5.53107261658

ur5_g_kdl= 12.5441551208

ur5_C_kdl= 15.8885717392

ur5_M_kdl= 14.0699148178
drbdl= 13.0543708801

fdcrba= 39.7365093231

fdaba= 37.5424623489


In [ ]:

ur5_g_u2c= 16.585111618

ur5_id_u2c= 36.719083786

ur5_C_u2c= 26.3614654541

ur5_M_u2c= 16.522526741

fdcrba= 39.9143695831

fdaba= 37.4803543091




ur5_g_rbdl= 6.27994537354

ur5_id_rbdl= 12.0129585266

ur5_C_rbdl= 8.66198539734

ur5_M_rbdl= 5.52153587341

ur5_g_pb= 5.00798225403

ur5_id_pb= 10.1070404053

ur5_C_pb= 7.7269077301

ur5_M_pb= 5.52296638489

ur5_g_kdl= 12.5441551208

ur5_C_kdl= 15.8885717392

ur5_M_kdl= 13.5569572449

fdrbdl= 13.2150650024


# Derivatives

## q

In [35]:
import casadi as cs

qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

C_j = cs.jacobian(c_u2c(qs, qsdot), qs)
g_j = cs.jacobian(g_u2c(qs), qs)
M_j = cs.jacobian(M_u2c(qs), qs)
invd_j = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qs)
fd_aba_j = cs.jacobian(aba_u2c(qs, qsdot, taus), qs)
fd_crba_j = cs.jacobian(crba_u2c(qs, qsdot, taus), qs)


g_j = cs.Function("g_j", [qs], [g_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
C_j = cs.Function("C_j", [qs, qsdot], [C_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
M_j = cs.Function("M_j", [qs], [M_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
invd_j = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_j = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_j = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [36]:
def fdcrba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_j(q_none, qdot_none, tau_none)

def fdaba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_j(q_none, qdot_none, tau_none)
    
    
def M_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2   
    M_j(q_none)
    
def g_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_j(q_none)


def C_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    C_j(q_none, qdot_none)
    
def ID_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    invd_j(q_none, qdot_none, qddot_none)

def aba_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_j(q_none, qdot_none, tau_none)
    
def crba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_j(q_none, qdot_none, tau_none)

In [40]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_gj = repeat("g_j_func()", setup = "from __main__ import g_j_func", repeat = nrepeat, number = nitr)
mediantime_gj = median(timeit_gj)/nitr*1000000

timeit_Cj = repeat("C_j_func()", setup = "from __main__ import C_j_func", repeat = nrepeat, number = nitr)
mediantime_Cj = median(timeit_Cj)/nitr*1000000

timeit_Mj = repeat("M_j_func()", setup = "from __main__ import M_j_func", repeat = nrepeat, number = nitr)
mediantime_Mj = median(timeit_Mj)/nitr*1000000

timeit_fdabaj = repeat("fdaba_j_func()", setup = "from __main__ import fdaba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("fdcrba_j_func()", setup = "from __main__ import fdcrba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_g_d=", mediantime_gj
print "\nur5_C_d=", mediantime_Cj
print "\nur5_M_d=", mediantime_Mj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 38.9980077744

ur5_g_d= 16.3205862045

ur5_C_d= 29.0064811707

ur5_M_d= 18.0535316467

ur5_fdaba_d= 46.2439060211

ur5_fdcrba_d= 49.4735240936




ur5_id_d= 39.0620231628

ur5_g_d= 16.2180662155

ur5_C_d= 29.2503833771

ur5_M_d= 18.0971622467

ur5_fdaba_d= 46.168088913

ur5_fdcrba_d= 49.1405725479






ur5_id_d= 36.6973876953

ur5_g_d= 16.0750150681

ur5_C_d= 28.6685228348

ur5_M_d= 17.9800987244

ur5_fdaba_d= 37.5105142593

ur5_fdcrba_d= 37.5329256058




## qdot

In [41]:
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

C_j = cs.jacobian(c_u2c(qs, qsdot), qsdot)
invd_j = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qsdot)
fd_aba_j = cs.jacobian(aba_u2c(qs, qsdot, taus), qsdot)
fd_crba_j = cs.jacobian(crba_u2c(qs, qsdot, taus), qsdot)


C_j = cs.Function("C_j", [qs, qsdot], [C_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
invd_j = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_j = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_j = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [43]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_Cj = repeat("C_j_func()", setup = "from __main__ import C_j_func", repeat = nrepeat, number = nitr)
mediantime_Cj = median(timeit_Cj)/nitr*1000000

timeit_fdabaj = repeat("fdaba_j_func()", setup = "from __main__ import fdaba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("fdcrba_j_func()", setup = "from __main__ import fdcrba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_C_d=", mediantime_Cj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 38.4650230408

ur5_C_d= 28.8100242615

ur5_fdaba_d= 41.4135456085

ur5_fdcrba_d= 40.058016777




ur5_id_d= 38.4650230408

ur5_C_d= 28.8100242615

ur5_fdaba_d= 41.2844419479

ur5_fdcrba_d= 39.7979021072


## qddot / tau


In [44]:
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

invd_j = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qsddot)
fd_aba_j = cs.jacobian(aba_u2c(qs, qsdot, taus), taus)
fd_crba_j = cs.jacobian(crba_u2c(qs, qsdot, taus), taus)

invd_j = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_j = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_j = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [46]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_fdabaj = repeat("fdaba_j_func()", setup = "from __main__ import fdaba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("fdcrba_j_func()", setup = "from __main__ import fdcrba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\nur5_id_d=", mediantime_idj
print "\nur5_fdaba_d=", mediantime_fd_abaj
print "\nur5_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 37.2290611267

ur5_fdaba_d= 37.9205942154

ur5_fdcrba_d= 37.5564098358


In [17]:

ur5_id_d= 37.1505022049

ur5_fdaba_d= 37.5800132751

ur5_fdcrba_d= 37.2189283371
